### ISSUE TITLE
Nested INSERT bug

### ISSUE THREAD

**Issue by vpetrovykh** [edgedb/edgedb#1080]
```
db> CREATE TYPE Foo {CREATE LINK bar -> Object};
CREATE
db> INSERT Foo {bar := (INSERT Foo)};
QueryError: invalid reference to default::Foo: self-referencing INSERTs are not allowed
Hint: Use DETACHED if you meant to refer to an uncorrelated default::Foo set
### INSERT Foo {bar := (INSERT Foo)};
###
```
This looks like a bug to me. I was under the impression that the thing right after `INSERT` is supposed to be a type name, not an arbitrary expression, as such there cannot be a problem of self-reference. If memory serves me, we couldn't actually guarantee that self-references in an `INSERT` would work (say for use in a subquery selecting some link based on other yet-to-be-inserted data that appears syntactically in the `INSERT` statement).

As far as I recall this is illegal:
```
INSERT Bar {
    a := 'data',
    b := .a
};
```
Am I misunderstanding `INSERT`?

Just to clarify, the intent is to simple create 2 `Foo` objects where one is linked to the other. This is *not* intended or expected to be some sort of self-link.

**Comment by elprans** [comment]
We prohibited this sort of self-reference for consistency with `UPDATE Foo`, where `Foo` *is* an expression.

**Comment by elprans** [comment]
See #539 for context of this.

**Comment by 1st1** [comment]
TBH, I still don't understand why `INSERT Foo {bar := (INSERT Foo)};` doesn't work.   This looks like something obvious that doesn't work for some non-obvious reasons -- people will struggle with this.

> UPDATE Foo, where Foo is an expression.

Can "Foo" be an expression?

**Comment by elprans** [comment]
> Can "Foo" be an expression?

I don't see why not.

```
WITH Foo := (SELECT User FILTER ...)
UPDATE Foo SET { ... };
```

`DELETE` is similar.

**Comment by 1st1** [comment]
DELETE requires parens, no? I.e. `DELETE (SELECT ...)`.

`INSERT Foo {bar := (INSERT Foo)};` seems more useful to me than `WITH Foo := (SELECT User FILTER ...) UPDATE Foo SET { ... };`.

**Comment by vpetrovykh** [comment]
OK, so I understand why we might want to forbid this:
```
INSERT Foo {bar := (SELECT Foo LIMIT 1)};
```
This is genuinely ambiguous in terms of what the person might be trying to express:
1) The intent was to create a self-looping single object `Foo`.
2) The intent was to actually create a new `Foo` with some other random existing `Foo` linked to it.

But the two `INSERT`s are not ambiguous in the same way - there's no alternative interpretation where "inserting the same thing twice" means "oh, just insert one of them and ignore the other".

**Comment by vpetrovykh** [comment]
A different way of thinking about this is that `SELECT`, `UPDATE`, and `DELETE` can only possibly refer to objects existing in the database, therefore they would act on expressions that denote those objects. However, `INSERT` specifically is creating a new thing that doesn't already exist, so the "name" to be inserted _cannot possibly_ be an expression of an existing thing, but has to be the _type_ of a thing to be created. I think it makes sense to treat `INSERT Foo` different from `SELECT Foo` or `DELETE Foo` because of that, despite the superficial syntactical similarity.

Also, would this also be an error (it happens to currently work)?
```
INSERT Foo {bar := <Foo>{}};
``` 
I'm aware that it's redundant, but is it illegal or ambiguous?

**Comment by Fogapod** [comment]
```edgeql
SELECT (
    INSERT User {nickname := 'bob', email := 'bob@example.com', password := b'hashed'}
    IF true
    ELSE 1
);```
Produces the same error. Is it related?

**Comment by elprans** [comment]
@Fogapod, `INSERT` cannot be used inside a conditional in general.  We're working on allowing it with `??` in certain cases (see #684).

**Comment by elprans** [comment]
> Also, would this also be an error (it happens to currently work)?
> 
> ```
> INSERT Foo {bar := <Foo>{}};
> ```
> 
> I'm aware that it's redundant, but is it illegal or ambiguous?

No, a bound variable reference and a type reference in a type expression are not related in any way.

I'm closing this as not a bug.

**Comment by elprans** [comment]
I'm reopening this for reconsideration, since there is currently no clean way of `INSERT`-ing self-referencing objects.  To improve the situation we have two options:

1) Allow `INSERT Foo {bar := (INSERT Foo)};` without qualifiers, since, as @vpetrovykh rightly pointed out in the [comment above](https://github.com/edgedb/edgedb/issues/1080#issuecomment-573282685), there is no ambiguity as to how to interpret the expression, it must always insert two distinct objects.

2) Allow and require `INSERT DETACHED Foo` in the nested `INSERT`.  The only benefit of this approach is visual consistency with other statements that already require `DETACHED` to refer to `Foo` in the body of its `INSERT`, most importantly the `UPDATE` statement.

I now lean toward option 1), as it's the least verbose option, and visual inconsistency with `UPDATE` can be explained in the error message and the documentation, whereas forcing the `DISTINCT` noise word is a price you'll need to pay every time.

@1st1, thoughts?
